src/index.d.ts

```diff
@@ -0,0 +1,50 @@
+/// <reference types="react" />
+
+declare module 'oy-vey' {
+  export const Table: React.DetailedHTMLFactory<Oy.OyTableElements, HTMLTableElement>;
+  export const TBody: React.DetailedHTMLFactory<Oy.OyTBodyElements, HTMLTableSectionElement>;
+  export const TR: React.DetailedHTMLFactory<Oy.OyTRElements, HTMLTableRowElement>;
+  export const TD: React.DetailedHTMLFactory<Oy.OyTDElements, HTMLTableDataCellElement>;
+  export const Img: React.DetailedHTMLFactory<Oy.OyImgElements, HTMLImageElement>;
+  export const A: React.DetailedHTMLFactory<Oy.OyAElements, HTMLAnchorElement>;
+  export const renderTemplate: (jsx: JSX.Element, opts: Oy.RenderOptions) => string;
+
+  const oy: {
+    Table: typeof Table;
+    TBody: typeof TBody;
+    TR: typeof TR;
+    TD: typeof TD;
+    Img: typeof Img;
+    A: typeof A;
+    renderTemplate: typeof renderTemplate;
+  };
+  export default oy;
+}
+
+declare namespace Oy {
```

Comment:
I don't think this needs to be declared?

Reply:
Bit of personal preference here. `declare`d namespaces, as opposed to exported namespaces or interfaces, form an ambient context, which makes it easier for users to extend them because they don't have to be concretely imported.

E.g. if I had this file in my workspace, and something was incorrect or out of date, then I could also make another file `oy-augmented.d.ts` with

```
declare namespace Oy {
  interface OyElement {
    newPropAddedinVersionX: string
  }
}
```

and Typescript would infer that an OyElement contains both declared sets of properties, without the user having to explicitly import the type from either file.

Reply:
`declare`d items mean that the author needs to ensure they would be available at runtime, no? Unlike, say jQuery, nothing in the library is designed be available in the ambient context—the used parts are to be imported explicitly.

w.r.t. merging, I don't think you need to declare a namespace to achieve declaration merging: "Similarly to interfaces, namespaces of the same name will also merge their members." https://www.typescriptlang.org/docs/handbook/declaration-merging.html If I'm interpreting the docs correctly, I imagine something like this would work:

```
// oy-augmented.d.ts
namespace Oy {
  interface OyElement {
    newPropAddedinVersionX: string;
  }
}
```

Reply:
>`declare`d items mean that the author needs to ensure they would be available at runtime, no? Unlike, say jQuery, nothing in the library is designed be available in the ambient context—the used parts are to be imported explicitly.

Yep, so `declare`d items mean that we're suggesting that something exists within the corresponding Javascript. If you declare an object, e.g. `declare var x: number`, then you need to ensure that `x` exists at runtime. This is what the `declare module 'oy-vey'` part is about - declaring all the _runtime objects_ that are provided by the author.

But the namespace defines _only_ types, not object instances, so all of it will be type-erased before runtime. This isn't a problem because it's not suggesting that any _object_ will exist at runtime. It just means users can reference the types in their own code, without affecting their runtime code.

E.g.
```
// InvoiceMailer.ts
import { renderTemplate, Table } from 'oy-vey'

const InvoiceTemplate = <Table>...,</Table>

export function sendInvoiceWithOpts(renderOptions: Oy.RenderOptions) {
  renderTemplate(InvoiceTemplate, renderOptions)
}
```

In the above, the user can say that it's function arguments are typed as `Oy.RenderOptions` without having to import that type, because it exists in an ambient context. And that's good. It doesn't affect runtime code, because all types are erased during compile. In fact, if the user _did_ have to import the type, then that _would_ cause runtime issues, because an import statement would not be type-erased during compilation.


>w.r.t. merging, I don't think you need to declare a namespace to achieve declaration merging: "Similarly to interfaces, namespaces of the same name will also merge their members." 

Shorter answer here: in a type declaration files (.d.ts), everything at the top-level needs to be `declare`d. Removing the `declare` modifier as you suggested will fail to compile.

Reply:
Got it, thanks for explaining! Other than module augmentation vs. namespace merging, are there additional differences for the user between:

```
declare module 'oy-vey' {
  ...
  namespace Oy { ... }
}
```

and what you have now? I'm a little worried about user confusion between `oy` and `Oy`, so I like the idea of making type extensions less obvious.

Reply:
Hmm, from my experience I don't think there's any other difference. In both cases, the types would appear under the `Oy` namespace and usage should be the same.